### PR TITLE
ssh/tailssh: try launching commands with /usr/bin/login on macOS

### PR DESCRIPTION
**Pull request by maisem**
Updates #4939

### CODE REVIEW

File: ssh/tailssh/incubator.go [hunk 1]
```diff
@@ -699,9 +699,39 @@ func fileExists(path string) bool {
...
+// The login binary is only used:
+//   - on darwin, if the client is requesting a shell or a command.
+//   - on linux and bsd, if the client is requesting a shell with a tty.
 func (ia *incubatorArgs) loginArgs() []string {
+	if ia.isSFTP {
+		return nil
+	}
 	switch runtime.GOOS {
+	case "darwin":
+		args := []string{ia.loginCmdPath,
```
Comment (bradfitz) [inline ssh/tailssh/incubator.go:713]: {\n

File: ssh/tailssh/incubator.go [hunk 2]
```diff
@@ -699,9 +699,39 @@ func fileExists(path string) bool {
 	return err == nil
 }
 
+// loginArgs returns the arguments to use to exec the login binary.
+// It returns nil if the login binary should not be used.
+// The login binary is only used:
+//   - on darwin, if the client is requesting a shell or a command.
+//   - on linux and bsd, if the client is requesting a shell with a tty.
```
Comment (bradfitz) [inline ssh/tailssh/incubator.go:706]: BSD .... TTY

File: ssh/tailssh/incubator.go [hunk 3]
```diff
@@ -124,10 +124,9 @@ func (ss *sshSession) newIncubatorCommand() (cmd *exec.Cmd) {
 	} else {
 		if isShell {
 			incubatorArgs = append(incubatorArgs, "--shell")
-			// Currently (2022-05-09) `login` is only used for shells
-			if lp, err := exec.LookPath("login"); err == nil {
-				incubatorArgs = append(incubatorArgs, "--login-cmd="+lp)
-			}
+		}
+		if lp, err := exec.LookPath("login"); err == nil {
+			incubatorArgs = append(incubatorArgs, "--login-cmd="+lp)
```
Comment (bradfitz) [inline ssh/tailssh/incubator.go:129]: commit message says "on macOS" but this doesn't look like macOS-specific code? you changing Linux behavior here?

Comment (maisem) [reply ssh/tailssh/incubator.go:129]: it adds a lookup but doesn't change behavior, will add an additional check

Comment (maisem) [reply ssh/tailssh/incubator.go]: Done

File: ssh/tailssh/incubator.go [hunk 4]
```diff
@@ -124,7 +124,8 @@ func (ss *sshSession) newIncubatorCommand() (cmd *exec.Cmd) {
 	} else {
 		if isShell {
 			incubatorArgs = append(incubatorArgs, "--shell")
-			// Currently (2022-05-09) `login` is only used for shells
+		}
+		if isShell || runtime.GOOS == "darwin" {
 			if lp, err := exec.LookPath("login"); err == nil {
 				incubatorArgs = append(incubatorArgs, "--login-cmd="+lp)
 			}
```
Comment (bradfitz) [inline ssh/tailssh/incubator.go:131]: add some comments about why darwin uses login for non-shell commands and why Linux etc don't?

File: ssh/tailssh/incubator.go [hunk 5]
```diff
@@ -699,9 +701,40 @@ func fileExists(path string) bool {
 	return err == nil
 }
 
+// loginArgs returns the arguments to use to exec the login binary.
+// It returns nil if the login binary should not be used.
```
Comment (bradfitz) [inline ssh/tailssh/incubator.go:705]: the caller on line 222 seems to care about non-empty length, not nil vs non-nil

make consistent one way or another? (docs & caller)

**Review by bradfitz — APPROVED** [review]

File: ssh/tailssh/incubator.go [hunk 6]
```diff
@@ -699,9 +701,40 @@ func fileExists(path string) bool {
...
 func (ia *incubatorArgs) loginArgs() []string {
+	if ia.isSFTP {
+		return nil
+	}
 	switch runtime.GOOS {
+	case "darwin":
+		args := []string{
+			ia.loginCmdPath,
+			"-f", // already authenticated
+			"-p", // discard previous environment, may change to -pq
```
Comment (Zaba) [inline ssh/tailssh/incubator.go:718]: The comment seems to indicate the intent is to discard the previous environment, but `-p` tells `login` to preserve it instead.

Comment (maisem) [reply ssh/tailssh/incubator.go:718]: sorry, the comment is wrong. I'll send a PR to update it. We want to preserve the environment that we passed in like `$SHELL` (everything from `envForUser`)

**Pull request merged**